src/utils/printing/getPrinterIppAttributes.ts

```diff
@@ -0,0 +1,218 @@
+import exec from '../exec'
+import { debug } from '.'
+import assert from 'assert'
+import tmp from 'tmp-promise'
+import fs from 'fs'
+
+/**
+ * IPP printer-state identifies the basic status of a printer.
+ * Spec: https://datatracker.ietf.org/doc/html/rfc2911#section-4.4.11
+ */
+export enum IppPrinterState {
+  Idle = 'idle', // 3
+  Processing = 'processing', // 4
+  Stopped = 'stopped', // 5
+}
+
+/**
+ * IPP printer-state-reasons explain what's going on with a printer in detail.
+ * Spec: https://datatracker.ietf.org/doc/html/rfc2911#section-4.4.12
+ * There are more possible reasons than covered in the spec, so we just type as string.
+ *
+ * Note that the actual printer-state-reasons sent by the printer may have a
+ * suffix of either: "-report", "-warning", or "-error" (e.g. "media-jam-error").
+ */
+export type IppPrinterStateReason = string
+
+/**
+ * "Marker" is a general name for ink/toner/etc. CUPS implements a variety of
+ * marker-related IPP attributes prefixed with "marker-", e.g. "marker-levels".
+ * Each attribute is a comma-delimated list of values, since a printer may have
+ * multiple marker supplies (e.g. black and color ink cartridges). Here, we
+ * represent the marker info for a single marker supply.
+ * Spec: https://www.cups.org/doc/spec-ipp.html
+ */
+export interface IppMarkerInfo {
+  name: string // e.g. "black cartridge"
+  color: string // e.g. "#000000"
+  type: string // e.g. "toner-cartridge"
+  lowLevel: number // e.g. 2
+  highLevel: number // e.g. 100
+  level: number // e.g. 83
+}
+
+/**
+ * A collection of status attributes we can get from a printer via IPP.
+ */
+export interface PrinterIppAttributes {
+  state: IppPrinterState | null
```

Comment:
No strong reason, just a different convention. Happy to switch it

Reply:
I guess the reason I've preferred `null` over `undefined` in the past is so that you are guaranteed to have the field still present. I find it confusing when fields disappear from objects. But I think with TS and good IDE support it's not as big of an issue. Still the JSON can be confusing though.

Reply:
So, following the idea of making fields undefined to represent the "didn't get any result" state, here are the types I came up with:

```ts
export type PrinterInfo = Omit<Electron.PrinterInfo, 'status'> &
  (PrinterIppAttributes | {}) & {
    connected: boolean
  }

export interface PrinterIppAttributes {
  state: IppPrinterState
  stateReasons: IppPrinterStateReason[]
  markerInfos: IppMarkerInfo[]
}
```
It's a bit more expressive, since we either get data for all three fields or we don't.

Does this seem more aligned to you? I still have a bit of hesitancy about disappearing fields but I'm willing to try it out.